### PR TITLE
nanocoap/sock: fix state handling for mismatching message ID in `nanocoap_sock_request_cb()`

### PR DESCRIPTION
**Pull request by fabian18**
### Contribution description

I encountered a situation in `nanocoap_sock_request_cb()` where the message ID did not match and an
`assert(ctx == NULL)` fired.
This introduces the state `STATE_WAIT_RESPONSE` after every (re)transmission.
The function is still hard to read. I added a label at the end to make sure the context is deleted in all cases.

### Issues/PRs references

I must have introduced this with #20697

### CODE REVIEW

File: sys/net/application_layer/nanocoap/sock.c [hunk 1]
```diff
@@ -315,24 +311,38 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
...
-                break;
+                continue;
             }
+        case STATE_RESPONSE_OK:             /* response received, release ctx */
+            goto release;
         }
     }
 
+release:
+    while(ctx) {
```
Comment (mguetschow) [inline sys/net/application_layer/nanocoap/sock.c:341]:
```suggestion
    while (ctx) {
```

File: sys/net/application_layer/nanocoap/sock.c [hunk 2]
```diff
@@ -315,24 +311,38 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
...
+            case COAP_TYPE_RST:
             case COAP_TYPE_CON:
             case COAP_TYPE_NON:
                 state = STATE_RESPONSE_OK;
-                if (coap_get_type(pkt) == COAP_TYPE_CON) {
-                    /* send ACK */
-                    _send_ack(sock, pkt);
+                if (coap_get_type(pkt) == COAP_TYPE_RST) {
+                     /* think about whether cb should be called on RST as well */
+                    res = -EBADMSG;
```
Comment (mguetschow) [inline sys/net/application_layer/nanocoap/sock.c:320]:
```suggestion
                    res = -EBADMSG;
                    continue;
```

File: sys/net/application_layer/nanocoap/sock.c [hunk 3]
```diff
@@ -315,24 +311,38 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
...
+                else {
+                    if (coap_get_type(pkt) == COAP_TYPE_CON) {
+                        _send_ack(sock, pkt);
+                    }
+                    if (cb) {
+                        res = cb(arg, pkt);
+                    }
+                    else {
+                        res = _get_error(pkt);
+                    }
```
Comment (mguetschow) [inline sys/net/application_layer/nanocoap/sock.c:331]:
```suggestion
              if (coap_get_type(pkt) == COAP_TYPE_CON) {
                  _send_ack(sock, pkt);
              }
              if (cb) {
                  res = cb(arg, pkt);
              }
              else {
                  res = _get_error(pkt);
```

together with the above suggestion

File: sys/net/application_layer/nanocoap/sock.c [hunk 4]
```diff
@@ -315,24 +311,38 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
...
                 }
-                break;
+                continue;
             }
+        case STATE_RESPONSE_OK:             /* response received, release ctx */
+            goto release;
```
Comment (benpicco) [inline sys/net/application_layer/nanocoap/sock.c:336]:
```suggestion
                goto release;
            }
```

I guess we can get rid of `STATE_RESPONSE_OK` entirely.

**Review by benpicco — APPROVED** [review]
`tests/net/gcoap_fileserver` is now working again

**Pull request merged**